[#45] Set correct GCC CFLAGS when compile

normanmaurer opened the pull request:
Motivation:

By default autotools use -g when compile via gcc and so add global debug symbols. This makes the compiled file very big. Beside this no GCC optimization settings are used.

Modifications:

Use our own configure.ac file which allows us to not use -g flag and setting -O3 for optimized code.

Result:

Smaller file size due not include debug symbols and faster code due optimizations done by GCC.

normanmaurer commented:
@trustin please check... will do some benchmarks in the meantime

nmittler commented:
yikes :)

ejona86 commented:
-O3. Good idea! :-)

Note that -Werror is frequently looked down upon because different compiler versions have different warnings. So if the user is using a newer or older version of the compiler they frequently will be unable to compile, especially with -Wall. Different header files can also cause unexpected trouble. The general recommendation is that you specify -Werror in your own personal CFLAGS.

trustin commented:
LGTM. Please address @ejona86's comment before cherry-picking though.

normanmaurer commented:
Addressed comment and cherry-picked as 14037dec78dd8bb8f7b1261d495f7505bbaa4560 into 1.1.33. 

Did a few benchmarks and found it slightly faster, but not much.
